File: typescript/src/skills/payments/settle-result.ts
/**
 * Reading `POST /api/payments/settle` (2026-09-18).
 *
 * WHY. Since the portal's S-148 fix `settlePaymentToken` charges a short
 * lock what it holds instead of nothing, and the route answers `success:
 * true` together with `partial`, and, when partial, `unbilled` and
 * `requested` (app/api/payments/settle/route.ts). The SDK callers read only
 * `success`, so a partial settle read as charged in full: the payment skill
 * marked the run paid, and an x402 settle reported a valid payment for an
 * amount it did not collect. Every caller now goes through this reader,
 * which keeps the platform's fields, logs a partial at warn, and leaves it to
 * the caller to report it as partial. The Python twin is
 * `webagents/agents/skills/robutler/payments/settle_result.py`.
 */

import type { PaymentSettleResult } from './types';

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

const numberOrUndefined = (v: unknown): number | undefined => (typeof v === 'number' && Number.isFinite(v) ? v : undefined);
const amountOrUndefined = (v: unknown): number | string | undefined =>
  typeof v === 'string' || (typeof v === 'number' && Number.isFinite(v)) ? v : undefined;

/** The settle answer as the SDK reads it; `where` names the caller in the warn line. */
export function readSettleResult(raw: unknown, where: string): PaymentSettleResult {
  const body = isRecord(raw) ? raw : {};
  const success = body.success === true;
  const out: PaymentSettleResult = { success, partial: success && body.partial === true };
  const charged = amountOrUndefined(body.charged);
  if (charged !== undefined) out.charged = charged;
  const remaining = amountOrUndefined(body.remaining);
  if (remaining !== undefined) out.remaining = remaining;
  for (const k of ['chargedDollars', 'unbilledDollars', 'requestedDollars', 'remainingDollars'] as const) {
    const v = numberOrUndefined(body[k]);
    if (v !== undefined) out[k] = v;
  }
  if (typeof body.unbilled === 'string') out.unbilled = body.unbilled;
  if (typeof body.requested === 'string') out.requested = body.requested;
  if (typeof body.error === 'string') out.error = body.error;
  if (out.partial) {
    console.warn(
      `[payments] ${where}: PARTIAL settle, charged ${out.chargedDollars ?? out.charged ?? '?'} of ` +
        `${out.requestedDollars ?? out.requested ?? '?'} (unbilled ${out.unbilledDollars ?? out.unbilled ?? '?'})`,
    );
  }
  return out;
}

File: python/webagents/agents/skills/robutler/payments/settle_result.py
"""
Reading `POST /api/payments/settle` (2026-09-18), the Python twin of the
TypeScript `src/skills/payments/settle-result.ts`.

WHY. Since the portal's S-148 fix `settlePaymentToken` charges a short lock
what it holds instead of nothing, and the route answers `success: true`
together with `partial`, and, when partial, `unbilled` and `requested`
(app/api/payments/settle/route.ts; `charged + unbilled = requested`, each in
nanocents as a string beside a `...Dollars` number). The SDK callers read only
`success`, so a partial settle read as charged in full. Every caller now goes
through this reader, which keeps the platform's fields, makes `partial` an
explicit bool (False on an older platform that never sends it, and never True
on a failure), logs a partial at warn, and leaves it to the caller to report
it as partial.
"""

from __future__ import annotations

import logging
from typing import Any, Dict, Mapping, Optional

_log = logging.getLogger(__name__)


def read_settle_result(raw: Any, where: str, logger: Optional[Any] = None) -> Dict[str, Any]:
    """The settle answer as the SDK reads it; `where` names the caller in the warn line."""
    body = dict(raw) if isinstance(raw, Mapping) else {}
    success = body.get("success") is True
    body["success"] = success
    body["partial"] = success and body.get("partial") is True
    if body["partial"]:
        charged = body.get("chargedDollars", body.get("charged", "?"))
        requested = body.get("requestedDollars", body.get("requested", "?"))
        unbilled = body.get("unbilledDollars", body.get("unbilled", "?"))
        (logger or _log).warning(f"[payments] {where}: PARTIAL settle, charged {charged} of {requested} (unbilled {unbilled})")
    return body


__all__ = ["read_settle_result"]
